### PR TITLE
Report total active threads

### PR DESCRIPTION
**Pull request by edman124**
These metrics are needed as a part of an internal project for thread scaling.
Total number of active threads will be joined with worker cpu utilization to see if we are bounded by the # of parallel work items, ultimately to reduce worker idleness.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by edman124** [comment]
ptal @twang126

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 1]
```diff
@@ -158,6 +158,7 @@ public class StreamingDataflowWorker {
 
   /** The idGenerator to generate unique id globally. */
   private static final IdGenerator idGenerator = IdGenerators.decrementingLongs();
+
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:161]: nit: undo

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 2]
```diff
@@ -255,7 +256,6 @@ private static MapTask parseMapTask(String input) throws IOException {
 
   public static void main(String[] args) throws Exception {
     JvmInitializers.runOnStartup();
-
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:258]: nit: undo

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 3]
```diff
@@ -996,7 +1003,7 @@ private void dispatchLoop() {
               inputDataWatermark,
               synchronizedProcessingTime,
               workItem,
-              /*getWorkStreamLatencies=*/ Collections.emptyList());
+              /* getWorkStreamLatencies= */ Collections.emptyList());
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1006]: nit: undo

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 4]
```diff
@@ -2277,7 +2288,8 @@ public void completeWork(ShardedKey shardedKey, long workToken) {
         if (completedWork.getWorkItem().getWorkToken() != workToken) {
           // Work may have been completed due to clearing of stuck commits.
           LOG.warn(
-              "Unable to complete due to token mismatch for key {} and token {}, actual token was {}.",
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2280]: nit: undo this change

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 5]
```diff
@@ -2356,7 +2368,8 @@ public void printActiveWork(PrintWriter writer) {
           "<table border=\"1\" "
               + "style=\"border-collapse:collapse;padding:5px;border-spacing:5px;border:1px\">");
       writer.println(
-          "<tr><th>Key</th><th>Token</th><th>Queued</th><th>Active For</th><th>State</th><th>State Active For</th></tr>");
+          "<tr><th>Key</th><th>Token</th><th>Queued</th><th>Active For</th><th>State</th><th>State"
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2371]: nit: undo

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 6]
```diff
@@ -273,7 +273,7 @@ private ParallelInstruction makeWindowingSourceInstruction(Coder<?> coder) {
         CloudObject.forClassName(
             "com.google.cloud.dataflow.sdk.util.TimerOrElement$TimerOrElementCoder");
     List<CloudObject> component =
-        Collections.singletonList(CloudObjects.asCloudObject(coder, /*sdkComponents=*/ null));
+        Collections.singletonList(CloudObjects.asCloudObject(coder, /* sdkComponents= */ null));
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:276]: nit: undo

**Review by twang126 — COMMENTED** [review]
pls take a quick look and undo the changes that are just formatting, etc.

**Comment by edman124** [comment]
Run Java PreCommit

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 7]
```diff
@@ -2923,6 +2942,90 @@ public void run() {
...
+          @Override
+          public void run() {
+            try {
+              Thread.sleep(2000);
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+            }
+          }
+        };
+    assertEquals(0, executor.activeCount());
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:3011]: is there a way to access the actual counter value? Looks like this is just capturing the executor's copy of the number of active threads? Ideally we'd look that the counter value also matches this

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 8]
```diff
@@ -2923,6 +2942,90 @@ public void run() {
     executor.shutdown();
   }
 
+  @Test
+  public void testActiveThreadMetric() throws Exception {
+    int maxThreads = 5;
+    int threadExpiration = 60;
```
Comment (twang126) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2948]: nit: add a unit to the name (seconds/ms/etc) so threadExpirationSecs or w/e

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java (hunk 7 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:3011]: I don't think so since we don't have a StreamingDataflowWorker object created, but I think getting the value from the executor is ok since all the logic is done in the executor.

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 9]
```diff
@@ -2923,6 +2942,90 @@ public void run() {
...
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+            }
+          }
+        };
+    assertEquals(0, executor.activeCount());
+
+    assertTrue(computationState.activateWork(key1Shard1, m1));
+    executor.execute(m1, m1.getWorkItem().getSerializedSize());
+    Thread.sleep(1000);
```
[the remark this answers is not in the document]
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java]: Yea the sleep is for making sure the thread is started. I've updated the test to wait for a notifcation on thread start for each workitem.

**Comment by edman124** [comment]
ptal @bvolpato

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 10]
```diff
@@ -615,6 +617,11 @@ public static StreamingDataflowWorker fromDataflowWorkerHarnessOptions(
     this.timeAtMaxActiveThreads =
         pendingCumulativeCounters.longSum(
             StreamingSystemCounterNames.TIME_AT_MAX_ACTIVE_THREADS.counterName());
+    this.activeThreads =
+        pendingCumulativeCounters.intSum(StreamingSystemCounterNames.ACTIVE_THREADS.counterName());
+    this.maxActiveThreads =
+        pendingCumulativeCounters.intSum(
```
Comment (bvolpato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:623]: intMax?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 11]
```diff
@@ -2027,6 +2034,10 @@ private void updateVMMetrics() {
   private void updateThreadMetrics() {
     timeAtMaxActiveThreads.getAndReset();
     timeAtMaxActiveThreads.addValue(workUnitExecutor.allThreadsActiveTime());
+    activeThreads.getAndReset();
+    activeThreads.addValue(workUnitExecutor.activeCount());
+    maxActiveThreads.getAndReset();
+    maxActiveThreads.addValue(chooseMaximumNumberOfThreads());
```
Comment (bvolpato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2040]: If you make above `intMax`, I believe here we should use `workUnitExecutor.activeCount()` as well?

Otherwise it's just a different metric for `numberOfWorkerHarnessThreads`? Not actual max utilized

**Review by bvolpato — CHANGES_REQUESTED** [review]
In general LGTM, shared suggestion to make max active threads more useful

**Comment by edman124** [comment]
ptal @jrmccluskey

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 10 above)
Comment (jrmccluskey) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:623]: +1

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 11 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2040]: We are ok if it is the same as numberOfWorkerHarnessThreads, we just want to have it exported as a counter, in case we dynamically adjust thread limit later on

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 10 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:623]: We want just max number threads (more in next comment)

**Comment by edman124** [comment]
ptal @damccorm should be ready to merge.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 11 above)
Comment (damccorm) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2040]: This also doesn't make sense to me. If `numberOfWorkerHarnessThreads` is good enough, why don't we just use that? If we eventually decide to adjust the thread limit, can't we react at that point?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 10 above)
Comment (bvolpato) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:623]: I don't think it makes sense to sum/aggregate thread # in general. This will monotonically increase for each call

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 12]
```diff
@@ -2923,6 +2943,109 @@ public void run() {
...
+        };
+    assertEquals(0, executor.activeCount());
+
+    assertTrue(computationState.activateWork(key1Shard1, m1));
+    synchronized (m1) {
+      executor.execute(m1, m1.getWorkItem().getSerializedSize());
+      LOG.info("[chengedward] waiting on thread start m1]");
+      m1.wait();
+    }
+    LOG.info("[chengedward] thread started m1]");
```
Comment (bvolpato) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:3029]: consider removing logs if they were mostly to debug, or let's at least remove personal information before sending this in

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 11 above)
Comment (twang126) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2040]: The counter name maxActiveThreads is misleading. Instead, I believe the intention is to measure the number of total allocated threads for the worker. We should probably rename the counter to something along the lines of totalAllocatedThreads.

In this case (maybe I'm wrong here) but I think the counter type (intSum) and value (chooseMaximumNumberOfThreads()) are actually what we want? The end goal is to be able to calculate the ratio between the total # of active threads across all user workers / the total # of threads that are allocated.

Comment (bvolpato) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2040]: I see. Yeah it's a bit confusing/misleading, but I guess you are just replicating what we have for javaHarnessUsedMemory/javaHarnessMaxMemory?

Giving it's cumulative, I feel like using a long makes more sense than int. For the default case of 300 threads, we could only collect a limited number of times before overflowing.

Today the default interval (`windmillHarnessUpdateReportingPeriod`) is 10s, so that would blow up after 2147483647/300/(60*60/10*24*365) = ~2 years.

Comment (twang126) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2040]: Yeah, it's exactly in the same spirit as javaHarnessUsedMemory/javaHarnessMaxMemory.

As for the long vs int counter type, I'm not familiar with how counters are reported, but I was somehow under the impression that with the GetAndReset, the value of this counter would never approach the int max limit? In that way, it's almost acting as a gauge.

Comment (bvolpato) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:2040]: You're right, it's always resetting. Makes sense.

**Review by bvolpato — APPROVED** [review]
LGTM, just get rid of some debugging / personal information if possible, please

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 11 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Changed the counter name to totalAllocatedThreads to better reflect our intentions.

**Review by twang126 — APPROVED** [review]

**Comment by bvolpato** [comment]
Please `./gradlew spotlessApply`

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 13]
```diff
@@ -2793,6 +2826,107 @@ public void testMaxThreadMetric() throws Exception {
...
+            "computation",
+            defaultMapTask(Arrays.asList(makeSourceInstruction(StringUtf8Coder.of()))),
+            executor,
+            ImmutableMap.of(),
+            null);
+
+    ShardedKey key1Shard1 = ShardedKey.create(ByteString.copyFromUtf8("key1"), 1);
+
+    // real work
+    MockActiveWork m1 =
```
Comment (damccorm) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2858]: The test is now failing to build, please address the precommit failures (e.g. https://github.com/apache/beam/actions/runs/6344373926/job/17234456430?pr=28513)

**Comment by damccorm** [comment]
Run Java PreCommit

**Review by damccorm — APPROVED** [review]
Thanks

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java [hunk 14]
```diff
@@ -2793,6 +2793,79 @@ public void testMaxThreadMetric() throws Exception {
...
+
+    Work m4 = createMockWork(4, sleepProcessWorkFn);
+    assertEquals(0, executor.activeCount());
+
+    assertTrue(computationState.activateWork(key1Shard1, m2));
+    synchronized (this) {
+      executor.execute(m2, m2.getWorkItem().getSerializedSize());
+      this.wait();
+      // Seems current executor executes the initial work item twice
+      this.wait();
```
Comment (Abacn) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2849]: this may have caused #28957

Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2849]: It could be the first three started runs before the synchronized block starts causing a deadlock. I have a fix in https://github.com/apache/beam/pull/29041 that uses countdown latches.

Comment (Abacn) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:2849]: Thanks for the fix. While #29041 is in progress, would you mind lgtm #29065 so the current stuck test won't affect the whole test suite so that it can still finish?
